Drop missed frames instead of trying to catch up

vbousquet opened the pull request:
This PR change the way the frames are dispatched to the destination ; instead of processing each frame the one after the other, it will process only the latest unprocessed frame, dropping any frame that may have been received while processing.

This change is needed to avoid a situation where the pinball controller process frames at a higher rate that the device can process. In this situation, the frames stack up and little by little, the display lags behind the controller. There has been a few post on VPForum on this or issue reports ([latest from Thalamus](https://www.vpforums.org/index.php?showtopic=43569&p=472062) that made me work on this).

I do not have hardware DMD but Thalamus did the testing and report that it improves (see [here](https://github.com/vbousquet/flexdmd/issues/21)). I also performed some testing on virtual DMD/video output and it does not have any side effect (as expected).

xantari commented:
Will this cause problems for coloring since it relies on frame matching to figure out how to apply the coloring?

vbousquet commented:
As far as I understand and I have tested, it should not break coloring since it is performed before dispatching the colored frame to the render targets.

To be more precise, coloring is performed [here](https://github.com/freezy/dmd-extensions/blob/master/LibDmd/RenderGraph.cs#L218) and [here](https://github.com/freezy/dmd-extensions/blob/master/LibDmd/RenderGraph.cs#L224) using direct Rx susbcription (```Do().Subscribe()```)

This PR modify the way the *colored frame* (so all the frames are sent to the converter and this PR modifies what happens afterward) is connected to the output which is done in a custom [```Subscribe``` method](https://github.com/freezy/dmd-extensions/blob/master/LibDmd/RenderGraph.cs#L866), itself exclusively called from the [```Connect``` method](https://github.com/freezy/dmd-extensions/blob/master/LibDmd/RenderGraph.cs#L473).

vbousquet commented:
Following Xantari comment, I dived again and performed some more testing and checking.
* This led me to found a bug I have introduced in my previous pushed commit (sorry for that): I have added the fix for this in the commit just above.
* I also found a line that feels wrong to me : https://github.com/freezy/dmd-extensions/blob/e7147ec04b790dc10b8a4df52cb24564d1b2395f/LibDmd/RenderGraph.cs#L796 I really think this should be ```TransformGray4``` since this is transforming a Gray4 frame. I did not find a way to test this conversion so I did not propose a commit since I can not be sure (is there a situation where this code is used ?).

vbousquet commented:
As there were request for downloading and testing on VPForums, here are the updated files to download and test:
* [DmdDevice.zip](https://github.com/freezy/dmd-extensions/files/6239095/DmdDevice.zip)
* [DmdDevice64.zip](https://github.com/freezy/dmd-extensions/files/6239097/DmdDevice64.zip)

Please, if you can do some testing with your PinDMD or whatever real DMD flavor you may have, especially with coloring (using palettes file xx.pal and xx.vni) or RGB animations (like in Metal Slug or Diablo tables), or tables that had lags or hickups on your DMD, and give us some feedback (even if it just works). This really helps since I can only do testing with the virtual DMD.

sverrewl commented:
I have tried many different things, even enabled at91jit on tables that I've had to put to disable before. xx.pal and xx.vni seems to work flawlessly. The only thing I've found to be broken is that the video seq of Diablo doesn't play. Black screen. Verified that my install is ok, downloaded table again and re-registered ultradmd and that works. We're talking pindmd3 still. Metal slug tested, no issues, Serious Sam II as well. I tried jumping back many different versions of dmddevice for the Diablo issue, but, was not able to get it to work on any.

freezy commented:
Awesome, thanks. 

I think you can change it from `TransformGray2` to `TransformGray4`. If you want to test it, make the Virtual DMD smaller (change `DmdWidth`/`DmdHeight`), force it non-resizable (implement `IFixedSizeDestination`), and try a SAM game. There are commented examples for all this I think.

vbousquet commented:
So, I made the test and it works with both TransformGray2 and TransformGray4 (using dmdext test with a coloredgray4 image and logging everything, to be sure the right code path was used). I still made the change (commit just above) for cleaner and logical code. I think this PR is good to merge. Regards !

freezy reviewed (approved):
Thanks again!
